Android: add the basic toolset activity with functionality that allows to unpack the files of the original game to the fheroes2 external data directory

oleg-derevenetz opened the pull request:
As well as download & extract the demo version of the original game.

Related to #6008
Close #6404

I'm not a very talented UI designer, so...

![Screenshot_20221231-102608](https://user-images.githubusercontent.com/32623900/210133509-f090d3e2-7050-4226-8d36-bb2ef5b36612.jpg)

Sort of "Quadratisch. Praktisch. Gut.", but it does its job. This also should resolve an issue with modern Android devices (AFAIK Android 13 suffers from it) when user just [doesn't have an access](https://github.com/ihhub/fheroes2/issues/6008#issuecomment-1291005787) to the `sdcard/Android/data/org.fheroes2` to copy game data files to it.

Plus fixed minor regression after #6382 - default fullscreen on Android was actually toggled only on the second launch, not on the first one.

Laserlicht commented:
> Sort of "Quadratisch. Praktisch. Gut."

If German is used here, then it should also be translated into German. 😉

File for: `android/app/src/main/res/values-de/strings.xml`

[german translation](https://github.com/ihhub/fheroes2/files/10310698/strings.zip)

oleg-derevenetz commented:
Hi @Laserlicht

> If German is used here, then it should also be translated into German. 😉

Added your translation. However there may still be at least claims to my wording, so this may not be the final version :)

> Maybe you could place the logo centered above the upper, long text. This is already in the resources anyway.

I added it, but it may be not visible on QVGA (320x427 dp) if there are both upper message and bottom progress indicator are visible at the same time, because there is no place for it. I solved this by allowing the logo to go through the top edge of the screen, so all important controls will still be visible.

ihhub commented:
Hi @oleg-derevenetz , I tested the changes on my phone and I think we might need to change the behavior of the launcher. I suggest to check whether directories (at least DATA) or even AGG files are present. If yes then we can do:
- immediately run the game
- show the launcher with a timer for 3-5 seconds and automatically launch the game if no button is pressed

We should add a button to download the demo. Knowing how people read even an obvious text I foresee some complains about it.

What do you think?

oleg-derevenetz commented:
Hi @ihhub this launcher already checks if at least `data/heroes2.agg` is present and if not, it doesn't allow to run the game at all. I'm against auto-run, because even if all necessary files are present, user may want to "upgrade" them from demo version to full version, or from SW to PoL, or just add a few maps (yes, it is possible too). There is a working hyperlink in the text describing why the game could not be run, user may just click on it and download the ZIP file with demo, and then unpack it using the "extract assets" button.

ihhub commented:
Hi @oleg-derevenetz , in this case I would suggest to at least add a time to startup if no action if being done. Another hypothetical (I am not sure how we can do this) is to add a button within the game to launch the launcher by closing the game.

oleg-derevenetz commented:
@ihhub do you think that user will patiently wait several seconds for a timer expiration when he can just click on the button and play the game immediately? :)

Laserlicht commented:
The easiest option would be a splashscreen (e.g. only the logo in fullscreen, and below a button to launch the "configurator"). Show this then 2-3 sec. If anything (except button) is touched on the splashscreen, then jump right into the game. If there are no assets, then jump straight into the configurator.

I mean currently you have to skip the intro-video by touch. Whether there is now a splashscreen comes before it also makes no longer so different I think.

oleg-derevenetz commented:
Replaced the hyperlink with the button (with the same action) and added an inactivity timer for 3 sec.

@Laserlicht

> I think the best option would be a splashscreen (e.g. only the logo in fullscreen, and below a button to launch the "configurator"). Show this then 2-3 sec. If anything (except button) is touched on the splashscreen, then jump right into the game. If there are no assets, then jump straight into the configurator.

This is basically how it works now, but there is just one launcher screen instead of two separate "splash screen with the button" and "configurator screen" activities. The difference would be that launcher activity has the logo and two buttons, and the splash screen has the logo and one button, so... it looks like the difference is negligible :)

oleg-derevenetz commented:
Also implemented some sort of inactivity timeout progress bar:

![Screenshot_20221228-235353](https://user-images.githubusercontent.com/32623900/209886604-f4bd3781-c59f-44e8-91a7-f6c928c88b1c.jpg)

ihhub commented:
Hi @oleg-derevenetz , what do you think about an option within the game to enable / disable the launcher? For example, by default it would be disabled by will be shown up since no resources are present. Then once a user downloads a demo version and starts the game for the next startup this launcher won't be shown up. However, for Android devices we can add a special option to enable it back (maybe some special file or something) then after exiting the real game we can still shown the launcher.

I'm just thinking how to make it more user-friendly as usually launchers are some kind of hacks in my opinion :)

oleg-derevenetz commented:
@ihhub

> I'm just thinking how to make it more user-friendly as usually launchers are some kind of hacks in my opinion :)

OK, here is another approach:

![screenshot](https://user-images.githubusercontent.com/32623900/209979401-4eaa2d26-1b69-4733-95e8-766e06a18153.jpg)

The idea is that there are two icons on the desktop instead of one. "fh2 Toolset" is essentially almost the same launcher (but without a timer), and "fheroes2" is the game activity. If there is no OG assets, "fheroes2" just starts the "fh2 Toolset". You can try this new approach in this PR.

oleg-derevenetz commented:
Hi @zenseii @Laserlicht @drevoborod what is your opinion regarding the current approach from this PR (apart from the too basic design of the toolset UI)?

ihhub commented:
Hi @oleg-derevenetz , I like this approach even more. However, I found that after running a build built in Android studio the app icon was replaced by `fh2 Toolset` at the main screen where my fheroes2 icon was :)

oleg-derevenetz commented:
@ihhub

> I like this approach even more. However, I found that after running a build built in Android studio the app icon was replaced by fh2 Toolset at the main screen where my fheroes2 icon was :)

Well, maybe that's because `ToolsetActivity` goes first in the app manifest (however I'm not sure about that). Anyway, this is a one-time issue :)

ihhub commented:
> @ihhub
> 
> > I like this approach even more. However, I found that after running a build built in Android studio the app icon was replaced by fh2 Toolset at the main screen where my fheroes2 icon was :)
> 
> Well, maybe that's because `ToolsetActivity` goes first in the app manifest (however I'm not sure about that). Anyway, this is a one-time issue :)

We should fix it to make sure that toolset is not installed on player's devices first :)

oleg-derevenetz commented:
@ihhub

> We should fix it to make sure that toolset is not installed on player's devices first :)

I'm not sure that there is 100% method to fix this. I can try to place the `GameActivity` first in the manifest.

oleg-derevenetz commented:
Yes, apparently switching helps - at least on simulator and on my other devices.

Laserlicht commented:
For me, the solution is also okay.

Another possibility for reentering the config screen would be [shortcuts](https://developer.android.com/develop/ui/views/launch/shortcuts). But they are afaik only supported from version 7.

oleg-derevenetz commented:
> Another possibility for reentering the config screen would be [shortcuts](https://developer.android.com/develop/ui/views/launch/shortcuts). But they are afaik only supported from version 7.

IMHO shortcuts are good when we want to have a way to start one `Activity` with the different parameters directly from the Android Launcher screen - for example we have one `EditorActivity` and we want to run it with the different document templates. Here we have just one task per `Activity` (`GameActivity` will run the game and `ToolsetActivity` will offer some tools), so "one icon per activity" approach looks more natural.

drevoborod commented:
> Hi @zenseii @Laserlicht @drevoborod what is your opinion regarding the current approach from this PR (apart from the too basic design of the toolset UI)?

For me, launcher which is explicitly run before the game starts is more preferrable decision. Also after first successful run it can be disabled by default but there should be a possibility to enable it again in game options. Corresponding message about such behaviour could be added to the launcher itself (something like "This screen will not appear again but it can be enabled in game configuration options".) I don't like unnecessary apps on my device because all of them by default are listed in same space and I need to group them manually.

oleg-derevenetz commented:
> For me, launcher which is explicitly executed before the game starts is more preferrable solution.

This toolset will be run be the main game activity automatically, but only if there are no assets from the original game. If there are already assets, then this additional screen on startup is really useless... unless you want to "upgrade" game assets (from demo version to full version or from Succession Wars to Price of Loyalty) or add a new maps (on Android 13 you can't just copy files to the `Android/data/org.fheroes2` subdir anymore). Launching this activity from the `fheroes2` UI is possible in theory, but we will need to have special Android-only in-game UI elements, what I would like to avoid. Let's leave Android-specific elements to the regular Android Launcher.

oleg-derevenetz commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -2,4 +2,14 @@
     <string name="app_name">fheroes2</string>
+    <string name="activity_toolset_label">fh2 Toolset</string>
+    <string name="activity_toolset_game_logo_img_content_description">fheroes2 logo</string>
+    <string name="activity_toolset_game_status_lbl_text">You cannot run the game at the moment, because some data files of the original game are missing. Please put the contents of the directory with the original Heroes of Might and Magic II game into a ZIP archive, copy it to this device and use the \"Extract HoMM2 assets\" button. If you do not have a copy of the original game, you can download ...
```
@zenseii

> Obviously I should've made a zip archive by selecting all the single files within, but this is just to show what a possibly common user mistake can be when asking them to make a zip archive.

Yeah, this should be described better. Also it would be good to warn user if no files was extracted.

> Why does it need to be zipped? Wouldn't it be easier for a user to be able to simply copy their installation folder?

Because of two things:

1. The new Android feature called "Scoped Storage". In [newer Android versions](https://developer.android.com/about/versions/11/privacy/storage), app can't just traverse directories and files on the shared storage that it doesn't created. I can try to add the button that will ask the user to select the directory (using `ACTION_OPEN_DOCUMENT_TREE`), but there is ton of restrictions regarding this too - for example, you cannot choose something in certain directories (e.g. "Downloads" directory) this way.
2. There are numerous issues with copying SW map files to the device because of their MP2 extension ([this is a typical example](https://github.com/ihhub/fheroes2/issues/6404)).

Branikolog replied:
Hi, @oleg-derevenetz 
I've tested the build from this PR. Nice work! I have only a few minor remarks.
Firstly, when I press download demo I'm transfer to my browser, where installation initiates automatically and the last open page is show (obviously, not related to demo download process). Not sure it's my device working in such strange manner, it would be nice if fheroes2 toolset will be all time on my screen, or at least, return back to toolset after browser was opened.
Secondly, if the archive I'm trying to extract, is missing data files, nothing is changes, so user couldn't know, what is wrong. Could we, probably point in app text, what exact folders does it require? 
I suppose, not whole H2 original game is necessary to be archived, so some users with limited storage would be grateful for not copying unnecessary files.

oleg-derevenetz replied:
@Branikolog

> Not sure it's my device working in such strange manner, it would be nice if fheroes2 toolset will be all time on my screen, or at least, return back to toolset after browser was opened.

This button just opens the link to the `h2demo.zip` file in the default browser (or in some 3-rd party download manager, if user installed such and prefers it). Browser (or 3-rd party download manager) naturally has no idea about the toolset app. User should press the Android "Back" button to return to the toolset once download is complete.

> Could we, probably point in app text, what exact folders does it require?

You can offer the desired wording :)

Branikolog replied:
@oleg-derevenetz 

> This button just opens the link to the `h2demo.zip` file in the default browser (or in some 3-rd party download manager, if user installed such and prefers it). Browser (or 3-rd party download manager) naturally has no idea about the toolset app. User should press the Android "Back" button to return to the toolset once download is complete.
> 

Okay, I got it.

> You can offer the desired wording :)

You know, I'm not a pro in wording issues. :)
I'd simply add 4 folders names in brackets... 
`Put the contents of the directory with the original Heroes of Might and Magic II game (or just Data, Maps, Music, Anim folders) into a zip archive....`

oleg-derevenetz replied:
@Branikolog I implemented the more "tolerant" handling of ZIP archives proposed by @Laserlicht. So now you can pack the `Homm2 Gold` directory and it still will work. Also toolset will warn you if it was unable to find any assets in your archive.

Branikolog replied:
@oleg-derevenetz 

> @Branikolog I implemented the more "tolerant" handling of ZIP archives proposed by @Laserlicht. So now you can pack the `Homm2 Gold` directory and it still will work. Also toolset will warn you if it was unable to find any assets in your archive.

Great! I'll have a look once I have free minute. 👍

Laserlicht commented:
> I believe what I did was make a zip archive which had the structure of Homm2 Gold.zip/Homm2 Gold/"all files".

Maybe we should make it a bit more tolerant and allow subfolders in the ZIP. So go through the whole ZIP directory tree and search for a "DATA" (or "MAPS", etc.) folder and take it, no matter on which level.

Experience shows that such errors occur in practice. And we should make it as easy as possible for the user. Not everyone is an IT specialist. And we should avoid unnecessary problems or bad store ratings.

With the Flatpak, I have now converted it so that the GOG-EXE can also be taken (we had already discussed). And thus also circumvented problems [(see here)](https://www.reddit.com/r/HoMM/comments/zy152v/homm_ii_fheroes2_on_steam_deck_need_help/). But now it is very simple and convenient.

oleg-derevenetz commented:
@Laserlicht

> Maybe we should make it a bit more tolerant and allow subfolders in the ZIP. So go through the whole ZIP directory tree and search for a "DATA" (or "MAPS", etc.) folder and take it, no matter on which level.

Yes, it's a good point.

> Experience shows that such errors occur in practice. And we should make it as easy as possible for the user. Not everyone is an IT specialist. And we should avoid unnecessary problems or bad store ratings.

The `fheroes2`-based (old engine) "Free Heroes 2" in the Google Play doesn't really kiss users' asses very much - for example, it demands from the user to put OG data files to the SD card by hand, and it 's not even too clear where exactly (however it at least offers to download the demo version if user doesn't have the OG), and still has 4 star ratings :)

> With the Flatpak, I have now changed it so that the GOG-EXE can also be taken (we had already discussed). And thus also circumvented problems [(see here)](https://www.reddit.com/r/HoMM/comments/zy152v/homm_ii_fheroes2_on_steam_deck_need_help/). But now it is very simple and convenient.

There are million (well, not exactly, but there are some) issues with GOG-EXE. For example, it doesn't have `ANIM` directory included. On Windows and Linux I handle this by special scripts that take care of extracting the aminations from the CD image (non-ISO image BTW so I had to convert it to the ISO first so `bsdtar` was able to take care of it). Let's take it outside the scope of this PR for now.

Branikolog commented:
> and still has 4 star ratings :)

And zounds of complains regarding installation problems. :)

oleg-derevenetz commented:
@Branikolog

> And zounds of complains regarding installation problems. :)

Well yes, if I wasn't (rather faintly) familiar with development for Android, I probably wouldn't understand what the "app-data/net.sourceforge.fheroes2" means either.

Laserlicht commented:
Another little thing as a question: Shouldn't we actually switch to SVG icons for the Android version? That seems more common to me.

I saw earlier that there are already svg icons for fheroes2. That is why I came up with.

See here: https://github.com/PapirusDevelopmentTeam/papirus-icon-theme/issues/2742 and https://github.com/PapirusDevelopmentTeam/papirus-icon-theme/commit/2564948de46bd6b4887f5baf3d94ad1f4ca6208f

oleg-derevenetz commented:
Hi @Laserlicht

> Another little thing as a question: Shouldn't we actually switch to SVG icons for the Android version? That seems more common to me.

From which version SVG icons are supported? I know only about so-called [vector drawables](https://developer.android.com/develop/ui/views/graphics/vector-drawable-resources) which are in fact not SVGs, but some home-grown Google XML format. Some SVG can be converted to it, but not all. In either case, I propose to address this as a separate PR.

Laserlicht commented:
> Hi @Laserlicht
> 
> > Another little thing as a question: Shouldn't we actually switch to SVG icons for the Android version? That seems more common to me.
> 
> From which version SVG icons are supported? I know only about so-called [vector drawables](https://developer.android.com/develop/ui/views/graphics/vector-drawable-resources) which are in fact not SVGs, but some home-grown Google XML format. Some SVG can be converted to it, but not all. In either case, I propose to address this as a separate PR.

Unfortunately I am not the Android developer. I was actually quite sure that it would work. But apparently (after googling) bitmaps are still desired by Android...

ihhub commented:
Hi @oleg-derevenetz , I tested the latest version. The solution looks good but we need to do 2 things:
- rename the helper from `fh2 Toolset` to probably `fh2 setup`. This is just my personal opinion as I think **setup** is more user friendly
- we need to change the icon for the toolset. 2 identical icons are always confusing. I can provide an icon just for this case.

oleg-derevenetz commented:
@ihhub

> rename the helper from fh2 Toolset to probably fh2 setup. This is just my personal opinion as I think setup is more user friendly

Well, technically there is no problem, but I believe that "setup" does not quite reflect reality in terms of the possibilities of using this activity, especially in future. I have plans to extend its functionality later by adding the ability to extract and take away the save files for example. That's not quite a "setup", it's more a "toolset" really :)

> we need to change the icon for the toolset. 2 identical icons are always confusing. I can provide an icon just for this case.

No problem, just attach the ZIP file with icon(s) :)

ihhub commented on android/app/src/main/res/values/strings.xml:
```diff
@@ -2,4 +2,15 @@
+    <string name="activity_toolset_label">fh2 Toolset</string>
+    <string name="activity_toolset_game_logo_img_content_description">fheroes2 logo</string>
+    <string name="activity_toolset_game_status_lbl_text"><b>You cannot run the game at the moment, because some data files of the original game are missing.</b>\n\nPlease put the contents of the directory with the original Heroes of Might and Magic II game (or just <b>ANIM</b>, <b>DATA</b>, <b>MAPS</b> and <b>MUSIC</b> folders) into a ZIP archive, copy it to this device, press the \"<b>Extract Ho ...
+    <string name="activity_toolset_run_game_btn_name">Run fheroes2</string>
```
How about we name this button as **Start game**? This is way more familiar way for players.

ihhub commented on android/app/src/main/res/layout/activity_toolset.xml:
```diff
@@ -0,0 +1,86 @@
+            android:text="@string/activity_toolset_game_status_lbl_text"
+            android:visibility="gone" />
+
+        <Button
```
To have slightly better UX design I suggest to make bottom padding little bigger to separate this button from others.

oleg-derevenetz replied:
Doubled the bottom margin of this button.

ihhub commented:
Hi @oleg-derevenetz , this is the best what @Branikolog can do for now :)
[logo_toolset.zip](https://github.com/ihhub/fheroes2/files/10337435/logo_toolset.zip)

Do we need higher than 144x144 pixels images?

oleg-derevenetz commented:
@ihhub

> this is the best what @Branikolog can do for now :)

Replaced the toolset icon by this one.

> Do we need higher than 144x144 pixels resolution images?

No, this set is enough.

Branikolog commented:
Hi, @oleg-derevenetz .
How do you think, should we also mention official store (link, probably too), where the original Heroes 2 game could be purchased, during the installation process? So every new player understands, where the original game resources could be taken from.

oleg-derevenetz commented:
@Branikolog

> How do you think, should we also mention official store (link, probably too), where the original Heroes 2 game could be purchased, during the installation process? So every new player understands, where the original game resources could be taken from.

Technically we can add a working link to that long description text that is shown when there is no OG data found, but I do not know how appropriate it will be.

Branikolog commented:
@oleg-derevenetz 

> > How do you think, should we also mention official store (link, probably too), where the original Heroes 2 game could be purchased, during the installation process? So every new player understands, where the original game resources could be taken from.
> 
> Technically we can add a working link to that long description text that is shown when there is no OG data found, but I do not know how appropriate it will be.

Can we change the behaviour of "Download Demo" button to open submenu, where player can choose to select demo downloading or link to official store by two different buttons? We can split that long description we already have and place part regarding download process of demo or original resources to new submenu? The button, could be called "Get resources" or "Download resources" instead of "Download demo".

oleg-derevenetz commented:
@Branikolog

> Can we change the behaviour of "Download Demo" button to open submenu, where player can choose to select demo downloading or link to official store by two different buttons?

How do you imagine using these buttons? Suppose that user press the "GOG Store" button on his Android device, and what's next? Currently he will need the PC anyway to run the installer, and only then he will be able to create a ZIP file with OG data files. I propose an incremental approach with discussion of specific improvements before their implementation. This PR is primarily intended to resolve the issues with MP2 files and `Android/data` access on Android 12+ devices.

ihhub commented:
Hi @oleg-derevenetz and @Branikolog , adding a link to any store selling the game is out of the scope of this pull request. Also there are minimum 2 stores: GOG and Ubisoft. Let's leave it outside this PR.

Branikolog commented:
@oleg-derevenetz 
I was talking  rather of adding a hint for user, where he can get needed files, than introducing a one-click solution for installing the application, since people sometimes tend to ask, where they can get the full version. 
Anyway, I'm satisfied with current implementation.

ihhub reviewed (commented):
Hi @oleg-derevenetz , I left only one more comment. Could you please take a look at it?

oleg-derevenetz commented on android/app/src/main/java/org/fheroes2/HoMM2AssetManagement.java:
```diff
@@ -0,0 +1,126 @@
+    {
+        // It is allowed to extract only files located in these subdirectories
+        final Set<String> allowedSubdirNames = new HashSet<>();
+        allowedSubdirNames.add( "anim" );
```
Hi @ihhub

`heroes2/anim` definitely not, because it's the same as `anim` (with the current "tolerant" logic the shortest path wins, so `heroes2/anim` will turn into just `anim`). Regarding `anim2` - in which distributions is it used and where exactly is it located? Right in the root dir of the installation path?

ihhub commented:
@oleg-derevenetz , huge thanks for these improvements!

oleg-derevenetz commented:
Hi @ihhub after merging of this PR, project has [more than one Java file](https://docs.sonarqube.org/latest/analyzing-source-code/languages/java/) and now SonarQube demands to provide compiled classes for all Java files and [fails](https://github.com/ihhub/fheroes2/actions/runs/3842572586/jobs/6544128440). What should we do? There are two options:

1. Build this project as Android project for Sonar;
2. Turn off the Java analysis completely.

ihhub commented:
Hi @oleg-derevenetz , as a temporary solution if enabling analysis for Java takes time we need to disable analysis for it and then if we have bandwidth add the analysis. This will help us to proceed with the rest of work.
